torch/onnx/_internal/fx/function_dispatcher.py

```diff
@@ -3,306 +3,372 @@
-        # Skip decomposition for op_overload as long as that op_overload has a corresponding ONNX
-        # symbolic function.
-        if (
-            "torch._refs" in decomp_fn.__module__
-            or op_overload in _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE
+    1. Use the torch.ops name to find the function:
+        a. Check if the ATen overload exists.
+        b. If not found, check ATen overload=default.
+
+    2. Find the best match among all overloaded functions:
+        a. If the types match perfectly, select the function.
+        b. Otherwise, find the nearest one with matching score.
+    """
+
+    def __init__(self, registry: registration.OnnxRegistry, opset_version: int = 18):
+        """Initialize the Dispatcher.
+
+        Args:
+            registry: The registration registry.
+            opset_version: The model opset version.
+        """
+        self._registry = registry
+        self._opset_version = opset_version
+
+    @property
+    def opset_version(self) -> int:
+        """Get the model opset version."""
+        return self._opset_version
+
+    @property
+    def registry(self) -> registration.OnnxRegistry:
+        """Get the registration registry."""
+        return self._registry
+
+    @_beartype.beartype
+    def dispatch(
+        self,
+        node: torch.fx.Node,
+        onnx_args: Sequence,
+        onnx_kwargs: Mapping,
+    ) -> Union["onnxscript.OnnxFunction", "onnxscript.TracedOnnxFunction"]:
+        """Dispatches an ONNX function based on the given FX node, arguments, and keyword arguments.
+
+        Args:
+            node: The TorchFX node to dispatch the function for.
+            onnx_args: The arguments of the ONNX function.
+            onnx_kwargs: The keyword arguments of the ONNX function.
+
+        Returns:
+            Either an `onnxscript.OnnxFunction` or `onnxscript.TracedOnnxFunction` instance based on the dispatch algorithm.
+
+        Raises:
+            RuntimeError: If there are no overloaded functions available for the given FX node.
+        """
+        aten_name = self._get_aten_name(node)
+        function_overloads = self._get_function_overloads(node, aten_name)
+
+        overload_match_ranking: Dict[
+            Union[onnxscript.OnnxFunction, onnxscript.TracedOnnxFunction], int
+        ] = {}
+        # TODO: Cache the OpSchemaWrapper so we don't need to run the init logic everytime
+        for overload_func in function_overloads:
+            function_opschema = OpSchemaWrapper(overload_func.op_schema)
+            if function_opschema.perfect_match_inputs(onnx_args, onnx_kwargs):
+                # If the perfect match is found, return the function
+                return overload_func
+            overload_match_ranking[overload_func] = function_opschema.match_score
+        # TODO(titaiwang): Add diagnostic message saying not perfect match get the best match
+        return max(overload_match_ranking, key=overload_match_ranking.get)  # type: ignore[arg-type]
```

Comment:
I thought about your suggestion. I think even if we finally decide to do type promotion in dispatcher, I image it would be after we find an overload, which means that function should be called next to dispatch(), not inside it.